teps/000X-trigger-crd.md

```diff
@@ -0,0 +1,351 @@
+
+### Non-Goals
+
+<!--
+What is out of scope for this TEP?  Listing non-goals helps to focus discussion
+and make progress.
+-->
+How to create Multitenant EventListener resource with appropriate permissions won't be part of this proposal.
+This will be handled by operator or admin of the cluster.
+
+## Requirements
+
+<!--
+List the requirements for this TEP.
+-->
+1. Add a new CRD Trigger which will be extracted from Trigger definition in EventListener Spec.
+2. Bind this Trigger CRD to EventListener.
+3. Ability to run Trigger resource in a different namespace from EventListener namespace.
+
+## Proposal
+
+<!--
+This is where we get down to the specifics of what the proposal actually is.
+This should have enough detail that reviewers can understand exactly what
+you're proposing, but should not include things like API designs or
+implementation.  The "Design Details" section below is for the real
+nitty-gritty.
+-->
+EventListener CRD will be split into two - a Trigger CRD and a EventListener CRD.
+
+### Trigger CRD
+A single Trigger defines configuration for processing events i.e. it consists of TriggerBindings,
+a single TriggerTemplate, and optionally interceptors and a serviceAccount. 
+```
+apiVersion: v1alpha1
+kind: Trigger
+metadata:
+  name: my-repo-trigger
+spec:
+  serviceAccountName: "blah"
+  interceptors:
+    cel:
+      filter: "$(header.eventType == "push")"
+  bindings:
+    - name: key
+      value: "$(body.commit_id)"
+  template: "ref-to-my-template"
+ status:
+   address:
+      url: "el-my-svc.cluster.local" # could also be an IP address?
+   conditions:
+      status: True
+      message: "Bound to EventListener my-el"
+```
+
+#### Trigger Status
+Trigger status will report url of EventListener bound and EventListener information to 
+which it is bound.
+
+### EventListener CRD
+EventListeners expose an addressable "Sink" to which incoming events in the form of HTTP requests
+are directed and process those requests with configuration from bound Triggers. Boundnamespace acts 
+as selector based on namespace. All triggers in namespaces specified are selected can be served by
+EventListener.
+
+```
+apiVersion: v1alpha1
+kind: EventListener
+metadata:
+  name: my-repo-trigger
+spec:
+  serviceAccountName: "blah"
+  boundnamespace:
+      - foo-ns
+      - bar-ns
+```
+
+### Path based EventListener
+Inside EventListener, we will deduce Trigger to execute based on the path of the request URL.
+To refer to trigger resource foo in namespace bar, we would url ```/bar/foo```. Instead of 
+EventListener iterating through every Triggers that can be serve by EventListener, EventListener
+will directly process the Trigger refer by URL.
+
+### User Stories (optional)
+
+<!--
+Detail the things that people will be able to do if this TEP is implemented.
+Include as much detail as possible so that people can understand the "how" of
+the system.  The goal here is to make this feel real for users without getting
+bogged down.
+-->
```

Comment:
the user stories section isn't about how the proposal will be implemented but is instead about what users of this feature need to do, e.g.: https://www.mountaingoatsoftware.com/agile/user-stories#:~:text=User%20stories%20are%20short%2C%20simple,so%20that%20.

This would be a great place to explain more about the specifics of the problem you are solving, e.g. why you need to create resources in so many namespaces, how many namespaces you are dealing with, etc.

You're probably dealing with at least a couple of different users here which are relevant to the proposal:
- The operator / administrator of the cluster
- The user who wants to trigger the creation of resources in the cluster